File: src/main/java/bg/geist/util/Flags.java
package bg.geist.util;


public class Flags {
    public static int set(int flags, int bitmask) {
        return (flags | bitmask);
    }
    public static int clear(int flags, int bitmask) {
        return (flags & ~bitmask);
    }
    public static int toggle(int flags, int bitmask) {
        return (flags ^ bitmask);
    }
    public static boolean isTrue(int flags, int bitmask) {
        return ((flags & bitmask) > 0);
    }
    public static boolean isFalse(int flags, int bitmask) {
        return ((flags & bitmask) == 0);
    }
    /**
     * Check if flags count < 2 (IsPowerOfTwo)
     *
     * @param bits flags
     * @return TRUE for single flag or no flags
     */
    public static boolean isSingle(int bits) {
        return ((bits & (bits - 1)) == 0);
    }
    /**
     * Index to bitmask: 0, 1, 2, 3 ... 30  =>  1, 2, 4, 8, ... 1073741824
     *
     * @param index (max: ~30)
     * @return bitmask
     */
    public static int toBits(int index) {
        return (1 << index);
    }
    /**
     * Set flags by length
     *
     * @param length Number of flags
     * @return flags
     */
    public static int toFlags(int length) {
        int flags = 0;
        for (int i = 0; i < length; i++) {
            flags = (flags | (1 << i));
        }
        return flags;
    }
    /**
     * Array of booleans to flags: [true, false, true] => 5
     *
     * @param arr Array of booleans
     * @return flags
     */
    public static int toFlags(Boolean[] arr) {
        if (arr == null) {
            return 0;
        }
        int bits = 0;
        for (int i = 0; i < arr.length; i++) {
            if (arr[i]) {
                bits += (1 << i);
            }
        }
        return bits;
    }
    /**
     * Bits to index
     *
     * base 2: Math.log2(bits)
     * @return index (0 for bits=0)
     */
    public static int toIndex(int bits) {
        int log = 0;
        if((bits & 0xffff0000) != 0) {bits >>>= 16; log = 16;}
        if(bits >= 256) {bits >>>= 8; log += 8;}
        if(bits >= 16 ) {bits >>>= 4; log += 4;}
        if(bits >= 4  ) {bits >>>= 2; log += 2;}
        return (log + (bits >>> 1));
    }
}

File: src/main/resources/static/js/modules/utils/flags.js
class Flags {
    static set(flags, bitmask) {
        return (flags | bitmask)
    }
    static clear(flags, bitmask) {
        return (flags & ~bitmask)
    }
    static toggle(flags, bitmask) {
        return (flags ^ bitmask)
    }
    static isTrue(flags, bitmask) {
        return ((flags & bitmask) > 0)
    }
    static isFalse(flags, bitmask) {
        return ((flags & bitmask) === 0)
    }
    /**
     * Index to bitmask: 0, 1, 2, 3 ... 30  =>  1, 2, 4, 8, ... 1073741824
     * 
     * @param {number} index (max: ~30)
     * @returns {number} bits
     */
    static toBits(index) {
        return (1 << index)
    }
    /**
     * Array of booleans to flags: [true, false, true] => 5
     * 
     * @param {boolean[]} arr 
     * @returns {number} flags
     */
    static toNumber(arr) {
        return arr.reduce((acc, c, i) => {
            return acc + (c ? (1 << i) : 0)
        }, 0)
    }
    /**
     * Bits to index
     * 
     * base 2: Math.log2(bits) 
     * @returns: 0 for bits=0
     */
    static toIndex(bits) {
        let log = 0;
        if( ( bits & 0xffff0000 ) !== 0 ) { bits >>>= 16; log = 16 }
        if( bits >= 256 ) { bits >>>= 8; log += 8 }
        if( bits >= 16  ) { bits >>>= 4; log += 4 }
        if( bits >= 4   ) { bits >>>= 2; log += 2 }
        return (log + ( bits >>> 1 ))
    }
    /**
     * Flags to array of booleans: (5, 3) => [true, false, true]
     * 
     * @param {number} flags 
     * @param {number} length 
     * @returns {boolean[]} 
     */
    static toArray(flags, length) {
        return Array.from({length}, (v, i) => Flags.isTrue(flags, Flags.toBits(i)))
    }
    /**
     * Set flags by length 
     * 
     * @param {number} length 
     * @returns {number} flags
     */
    static byLength(length) {
        let flags = 0;
        for (let i = 0; i < length; i++) {
            flags = (flags | (1 << i))
        }
        return flags
    }
}

class ArrayOfFlags {
    constructor() {
        this.number = 0;
        this.length = 0;
    }
    get length() {
        return this._length
    }
    set length(length) {
        this._length = length;
        this.number = Flags.byLength(length);
    }
    getIndex(index) {
        return Flags.isTrue(this.number, Flags.toBits(index))
    }
    setIndex(index, flag) {
        this.number = flag ? Flags.set(this.number, Flags.toBits(index)) : Flags.clear(this.number, Flags.toBits(index))
    }
    toggle(index) {
        this.number = Flags.toggle(this.number, Flags.toBits(index))
    }
    get array() {
        return Flags.toArray(this.number, this.length)
    }
    set array(arr) {
        this._length = arr.length;
        this.number = Flags.toNumber(arr)
    }
}

export {Flags, ArrayOfFlags}